File: bin/build-docs.py
#!/usr/bin/env python

import json
import argparse

parser = argparse.ArgumentParser(description='Convert JSON docs to markdown.')
parser.add_argument('input', metavar='INPUT', type=str, help='The JSON source file.')

args = parser.parse_args()

inputPath = args.input
outputPath = '../docs/'

docs = {}

def getOutputPrefix(className):
	if isLibrary(className):
		return 'JS library_'
	if isWidget(className):
		return 'Widget_'

	return ''

def isGlobalObject(className):
	return className in ['Object', 'Function', 'Boolean', 'Error', 'Number', 'Date', 'String', 'RegExp', 'Array']

def isLibrary(className):
	return (className.find('Webos.') == 0)

def isWidget(className):
	return (className.find('$.webos.') == 0)

def isDocumented(className):
	return isLibrary(className) or isWidget(className) or isGlobalObject(className) or className in docs

def getOutputIndex(className):
	outputPrefix = getOutputPrefix(className)

	if isGlobalObject(className):
		return 'https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/'+className

	if isLibrary(className):
		className = className[len('Webos.'):]
	if isWidget(className):
		className = className[len('$.webos.'):]

	return outputPrefix+className.lower()

def getOutputPath(className):
	return getOutputIndex(className)+'.md'

def isPrivate(block):
	for tag in block['tags']:
		if tag['type'] == 'private':
			return True

	return False

def formatSince(tags):
	since = ''
	for tag in tags:
		if tag['type'] == 'since':
			since = tag['string']
			return '\n\nSince ['+since+'](../releases/tag/'+since+').'

	return ''

def formatAugments(tags):
	augments = ''
	for tag in tags:
		if tag['type'] == 'augments':
			augments = tag['otherClass']
			return '\n\nChild of '+formatTypes(augments)+'.'

	return ''

def formatInlineDeprecated(tags):
	deprecated = ''
	for tag in tags:
		if tag['type'] == 'deprecated':
			deprecated = tag['string']
			return '**Deprecated**: '+deprecated+'.'

	return ''

def formatDeprecated(tags):
	deprecated = formatInlineDeprecated(tags)

	if len(deprecated) > 0:
		return '\n\n'+deprecated
	else:
		return ''

def formatTypes(types):
	if type(types) == str:
		types = [types]
	
	i = 0
	for typeName in types:
		if typeName[0] == '{' and typeName[-1] == '}':
			typeName = typeName[1:-1]

		if isDocumented(typeName):
			types[i] = '['+typeName+']('+getOutputIndex(typeName)+')'
		else:
			types[i] = typeName
		i += 1
		
	return '_'+'|'.join(types)+'_'

def formatReturnTypes(tags):
	for tag in tags:
		if tag['type'] == 'return' or tag['type'] == 'returns':
			if 'types' in tag:
				return formatTypes(tag['types'])+' '
	return ''

def formatParamsList(tags):
	paramsList = []
	for tag in tags:
		if tag['type'] == 'param':
			if 'types' in tag:
				#paramTypes = formatTypes(tag['types']) #Do not print param types in params list
				paramTypes = ''
			else:
				paramTypes = ''
			paramsList.append(paramTypes+'**'+tag['name']+'**')
	return ', '.join(paramsList)

def formatParamsDescription(tags):
	paramsDescription = []
	for tag in tags:
		if tag['type'] == 'param':
			if 'types' in tag:
				paramTypes = formatTypes(tag['types'])+' '
			else:
				paramTypes = ''
			paramsDescription.append(paramTypes+'**'+tag['name']+'** '+tag['description'])
	if len(paramsDescription) == 0:
		return ''
	else:
		return '\n * ' + '\n * '.join(paramsDescription)

try:
	data = json.load(open(inputPath,'r'))
except ValueError:
	data = []

# Order by type
fileOverview = None
i = 0
for block in data:
	if 'ctx' in block:
		ctx = block['ctx']

		# Is it a class ?
		if ctx['type'] == 'constructor':
			if 'receiver' in ctx:
				className = ctx['receiver'] + '.' + ctx['name']
			else:
				className = ctx['name']

			if className in docs:
				docs[className]['class'] = block
			else:
				docs[className] = {
					'class': block,
					'static_method': [],
					'method': [],
					'property':[],
					'static_property':[],
					'prototype':[]
				}
			continue

		parent = None
		static = True
		if 'constructor' in ctx:
			parent = ctx['constructor']
			static = False
		elif 'receiver' in ctx:
			parent = ctx['receiver']

		if parent is None:
			continue

		if isPrivate(block):
			continue
		
		blockType = ctx['type']
		if static and blockType != 'prototype':
			blockType = 'static_'+blockType

		if parent not in docs:
			docs[parent] = {
				'class': None,
				'static_method': [],
				'method': [],
				'property':[],
				'static_property':[],
				'prototype':[]
			}
		docs[parent][blockType].append(block)
	else:
		if i == 0: # First tag & without ctx, maybe a file overview
			fileOverview = block
	i += 1


# For each class
for className in docs.keys():
	classDoc = docs[className]

	if className == 'this':
		continue

	output = ''

	# Class
	if 'class' in classDoc and classDoc['class'] is not None:
		output += classDoc['class']['description']['full']
		output += formatSince(classDoc['class']['tags'])
		output += formatAugments(classDoc['class']['tags'])
		output += formatDeprecated(classDoc['class']['tags'])
		output += '\n\n'

	# Options
	if isWidget(className):
		output += '# Options\n\n'

		options = None
		for block in classDoc['property']:
			ctx = block['ctx']
			if ctx['name'] == 'options':
				options = block
				break

		if options is not None:
			output += block['description']['full']+'\n'
		else:
			output += '_This widget hasn\'t any option._\n'

		output += '\n'
	else:
		pass #TODO: print properties

	# Methods
	output += '# Methods\n\n'
	for block in classDoc['method']:
		ctx = block['ctx']

		methodName = '**'+ctx['name']+'**'

		returnTypes = formatReturnTypes(block['tags'])
		paramsList = formatParamsList(block['tags'])
		paramsDescription = formatParamsDescription(block['tags'])
		deprecated = formatInlineDeprecated(block['tags'])
		summary = block['description']['summary'].strip()

		output += '* '+returnTypes+methodName+'('+paramsList+') : '+summary

		if len(deprecated) > 0:
			output += '\n  '+deprecated
		if len(paramsDescription) > 0:
			output += paramsDescription

		output += '\n'

	if len(classDoc['method']) == 0:
		output += '_This class hasn\'t any method._\n'

	output += '\n'

	# Static methods
	output += '# Static methods\n\n'
	for block in classDoc['static_method']:
		ctx = block['ctx']

		methodName = ctx['receiver'] + '.**'+ctx['name']+'**'

		returnTypes = formatReturnTypes(block['tags'])
		paramsList = formatParamsList(block['tags'])
		paramsDescription = formatParamsDescription(block['tags'])
		deprecated = formatInlineDeprecated(block['tags'])
		summary = block['description']['summary'].strip()

		output += '* '+returnTypes+methodName+'('+paramsList+') : '+summary

		if len(deprecated) > 0:
			output += '\n  '+deprecated
		if len(paramsDescription) > 0:
			output += paramsDescription

		output += '\n'

	if len(classDoc['static_method']) == 0:
		output += '_This class hasn\'t any static method._\n'

	outputFile = open(outputPath+'/'+getOutputPath(className), 'w')
	outputFile.write(output)

File: bin/build-docs.js
#!/usr/bin/env node

var fs = require('fs');
var dox = require('dox');

// Parse argv
var opt = require('node-getopt').create([
	['h' , 'help', 'display this help']
])
.bindHelp()
.parseSystem();

var inputPath = opt.argv[0],
	outputPath = '../docs/';

var docs = {};

function getOutputPrefix(className) {
	if (isLibrary(className)) {
		return 'JS library_';
	}
	if (isWidget(className)) {
		return 'Widget_';
	}

	return '';
}

function isGlobalObject(className) {
	return ~['Object', 'Function', 'Boolean', 'Error', 'Number', 'Date', 'String', 'RegExp', 'Array'].indexOf(className);
}

function isLibrary(className) {
	return (className.indexOf('Webos.') == 0);
}

function isWidget(className) {
	return (className.indexOf('$.webos.') == 0);
}

function isDocumented(className) {
	return (isLibrary(className) ||
		isWidget(className) ||
		isGlobalObject(className) ||
		docs[className]);
}

function getOutputIndex(className) {
	var outputPrefix = getOutputPrefix(className);

	if (isGlobalObject(className)) {
		return 'https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/'+className;
	}

	if (isLibrary(className)) {
		className = className.substr('Webos.'.length);
	}
	if (isWidget(className)) {
		className = className.substr('$.webos.'.length);
	}

	return outputPrefix + className.toLowerCase();
}

function getOutputPath(className) {
	return getOutputIndex(className)+'.md';
}

function isPrivate(block) {
	for (var i = 0; i < block.tags.length; i++) {
		var tag = block.tags[i];

		if (tag.type == 'private') {
			return true;
		}
	}

	return false;
}

function formatSince(tags) {
	for (var i = 0; i < tags.length; i++) {
		var tag = tags[i];

		if (tag.type == 'since') {
			var since = tag.string;
			return '\n\nSince ['+since+'](../releases/tag/'+since+').';
		}
	}

	return '';
}

function formatAugments(tags) {
	for (var i = 0; i < tags.length; i++) {
		var tag = tags[i];

		if (tag.type == 'augments') {
			var augments = tag.otherClass;
			return '\n\nChild of '+formatTypes(augments)+'.';
		}
	}

	return '';
}

function formatInlineDeprecated(tags) {
	for (var i = 0; i < tags.length; i++) {
		var tag = tags[i];

		if (tag.type == 'deprecated') {
			var deprecated = tag.string;

			if (deprecated) {
				return 'Deprecated: '+deprecated+'.';
			} else {
				return 'Deprecated.';
			}
		}
	}

	return '';
}

function formatDeprecated(tags) {
	var deprecated = formatInlineDeprecated(tags);

	if (deprecated) {
		return '\n\n'+deprecated;
	}

	return '';
}

function formatTypes(types) {
	if (typeof types == 'string') {
		types = [types];
	}

	for (var i = 0; i < types.length; i++) {
		var typeName = types[i];

		if (typeName[0] == '{' && typeName[typeName.length-1] == '}') {
			typeName = typeName.substr(1, typeName.length-2);
		}

		if (isDocumented(typeName)) {
			types[i] = '['+typeName+']('+getOutputIndex(typeName)+')';
		} else {
			types[i] = typeName;
		}
	}

	return '_'+types.join('|')+'_';
}

function formatReturnTypes(tags) {
	for (var i = 0; i < tags.length; i++) {
		var tag = tags[i];

		if (tag.type == 'return' || tag.type == 'returns') {
			if (tag.types) {
				return formatTypes(tag.types)+' ';
			}
		}
	}

	return '';
}

function formatParamsList(tags) {
	var paramsList = [];

	for (var i = 0; i < tags.length; i++) {
		var tag = tags[i];

		if (tag.type == 'param') {
			var paramTypes = '';
			if (tag.types) {
				//paramTypes = formatTypes(tag.types)+' '; // Do not print param types in params list
			}
			//paramsList.push(paramTypes+'**'+tag.name+'**');
			paramsList.push(paramTypes+tag.name);
		}
	}

	return paramsList.join(', ');
}

function formatParamsDescription(tags) {
	var paramsDescription = [];

	for (var i = 0; i < tags.length; i++) {
		var tag = tags[i];

		if (tag.type == 'param') {
			var paramTypes = '';
			if (tag.types) {
				paramTypes = formatTypes(tag.types)+' ';
			}
			paramsDescription.push(paramTypes+'**'+tag.name+'** '+tag.description);
		}
	}

	if (!paramsDescription.length) {
		return '';
	} else {

	}

	return '\n * ' + paramsDescription.join('\n * ');
}

function formatFile(data) {
	var fileOverview;

	// Order by type
	for (var i = 0; i < data.length; i++) {
		var block = data[i];

		if (block.ctx) {
			var ctx = block.ctx;

			// Is it a class ?
			if (ctx.type == 'constructor') {
				var className = ctx.name;
				if (ctx.receiver) {
					className = ctx.receiver + '.' + ctx.name
				}

				if (docs[className]) {
					docs[className].class = block;
				} else {
					docs[className] = {
						class: block,
						static_method: [],
						method: [],
						property: [],
						static_property: [],
						prototype: []
					};
					continue;
				}
			}

			var parent,
				isStatic = true;

			if (typeof ctx.constructor == 'string') {
				parent = ctx.constructor;
				isStatic = false;
			} else if (ctx.receiver) {
				parent = ctx.receiver;
			}

			if (!parent) {
				continue;
			}
			if (isPrivate(block)) {
				continue;
			}

			var blockType = ctx.type;
			if (isStatic && blockType != 'prototype') {
				blockType = 'static_'+blockType;
			}

			if (!docs[parent]) {
				docs[parent] = {
					class: null,
					static_method: [],
					method: [],
					property: [],
					static_property: [],
					prototype: []
				};
			}

			docs[parent][blockType].push(block)
		} else {
			if (i == 0) { // First tag & without ctx, maybe a file overview
				fileOverview = block;
			}
		}
	}

	// For each class
	var outputs = {};
	for (var className in docs) {
		var classDoc = docs[className];

		if (className == 'this') {
			continue;
		}

		var output = '';

		// Class
		if (classDoc.class) {
			output += classDoc['class']['description']['full'];
			output += formatSince(classDoc['class']['tags']);
			output += formatAugments(classDoc['class']['tags']);
			output += formatDeprecated(classDoc['class']['tags']);
			output += '\n\n';
		}

		// Options
		if (isWidget(className)) {
			output += '# Options\n\n';

			var options;
			for (var i = 0; i < classDoc.property.length; i++) {
				var block = classDoc.property[i], ctx = block.ctx;
				if (ctx.name == 'options') {
					options = block;
					break;
				}
			}
			if (options) {
				output += block['description']['full']+'\n';
			} else {
				output += '_This widget hasn\'t any option._\n';
			}
		} else {
			// TODO: print properties
		}

		// Methods
		output += '# Methods\n\n';
		for (var i = 0; i < classDoc.method.length; i++) {
			var block = classDoc.method[i], ctx = block.ctx;
			
			var methodName = '**'+ctx.name+'**';

			var returnTypes = formatReturnTypes(block['tags']),
				paramsList = formatParamsList(block['tags']),
				paramsDescription = formatParamsDescription(block['tags']),
				deprecated = formatInlineDeprecated(block['tags']),
				summary = block['description']['summary'].trim();

			output += '* '+methodName+'('+paramsList+') : '+summary;

			if (returnTypes) {
				output += '\n  Returns: '+returnTypes;
			}
			if (deprecated) {
				output += '\n  '+deprecated;
			}
			if (paramsDescription) {
				output += paramsDescription;
			}

			output += '\n';
		}
		if (!classDoc.method.length) {
			output += '_This class hasn\'t any method._\n';
		}

		output += '\n';

		// Static methods
		output += '# Static methods\n\n';
		for (var i = 0; i < classDoc.static_method.length; i++) {
			var block = classDoc.static_method[i], ctx = block.ctx;
			
			var methodName = ctx.receiver + '.**'+ctx.name+'**';

			var returnTypes = formatReturnTypes(block['tags']),
				paramsList = formatParamsList(block['tags']),
				paramsDescription = formatParamsDescription(block['tags']),
				deprecated = formatInlineDeprecated(block['tags']),
				summary = block['description']['summary'].trim();

			output += '* '+methodName+'('+paramsList+') : '+summary;

			if (returnTypes) {
				output += '\n  Returns: '+returnTypes;
			}
			if (deprecated) {
				output += '\n  '+deprecated;
			}
			if (paramsDescription) {
				output += paramsDescription;
			}

			output += '\n';
		}
		if (!classDoc.static_method.length) {
			output += '_This class hasn\'t any static method._\n';
		}

		outputs[className] = output;
	}

	return outputs;
}

fs.readFile(inputPath, function (err, buf) {
	if (err) {
		console.error(err);
		process.exit(1);
		return;
	}

	var data = dox.parseComments(buf.toString(), {
		raw: true
	});

	var outputs = formatFile(data);

	var completed = 0;
	function wrote(className) {
		completed++;
		console.log('Wrote docs for '+className);

		if (completed == outputs.length) {
			process.exit(0);
		}
	}
	function write(className, output) {
		fs.writeFile(outputPath+'/'+getOutputPath(className), output, function (err) {
			if (err) {
				console.error(err);
				process.exit(1);
				return;
			}

			wrote(className);
		});
	}

	for (var className in outputs) {
		write(className, outputs[className]);
	}
});
